File: 프로그래머스/lv2/84512. 모음 사전/모음 사전.java
import java.util.*;

class Solution {
    
    static ArrayList<String> list;
    static String[] vowels = {"A", "E", "I", "O", "U"};
    
    public int solution(String word) {
        list = new ArrayList<String>();
        String s = "";
        
        recur(s, 0);
        
        Collections.sort(list, new Comparator<>() {
            @Override
            public int compare(String s1, String s2) {
                return s1.compareTo(s2);
            }
        });
        
        int answer = list.indexOf(word) + 1;
        
        return answer;
    }
    
    static void recur(String word, int k) {
        if(k == 5) {
            if(word.equals("")) return;
            if(list.contains(word)) return;
            list.add(word);
            return;
        }
        
        for(int i = 0; i < 5; i++) {
            recur(word + vowels[i], k + 1);
            recur(word, k + 1);
        }
    }
}

File: 프로그래머스/2/84512. 모음 사전/모음 사전.js
function solution(word) {
    init('');
    
    list.sort();
    
    for(let i = 1; i <= list.length; i++) {
        if(list[i - 1] === word) {
            return i;
        }
    }
    
    return -1;
}

const vowel = ['A', 'E', 'I', 'O', 'U'];
let list = [];

const init = (str, idx) => {
    if(str.length > 5) return;
    
    if(str.length > 0 && !list.includes(str)) {
        list.push(str);
    }
    
    for(let i = 0; i < 5; i++) {
        init(str + vowel[i]);
    }
}